test/gtest/lattices.cpp

```diff
@@ -852,3 +855,499 @@ TEST(AbstractionLattice, Join) {
+  Element eqNull2;
+  Element eqNonNull2;
+  Element structNull1;
+  Element structNonNull1;
+
+  void checkJoin(const Element& a,
+                 const Element& b,
+                 const Element& join,
+                 const char* file,
+                 int line) {
+    testing::ScopedTrace trace(file, line, "check join");
+    Element copy = a;
+    EXPECT_EQ(lattice.join(copy, b), a != join);
+    EXPECT_EQ(copy, join);
+    copy = b;
+    EXPECT_EQ(lattice.join(copy, a), b != join);
+    EXPECT_EQ(copy, join);
+  }
+#define CHECK_JOIN(a, b, join) checkJoin(a, b, join, __FILE__, __LINE__)
+
+  void checkMeet(const Element& a,
+                 const Element& b,
+                 const Element& meet,
+                 const char* file,
+                 int line) {
+    testing::ScopedTrace trace(file, line, "check meet");
+    Element copy = a;
+    EXPECT_EQ(lattice.meet(copy, b), a != meet);
+    EXPECT_EQ(copy, meet);
+    copy = b;
+    EXPECT_EQ(lattice.meet(copy, a), b != meet);
+    EXPECT_EQ(copy, meet);
+  }
+#define CHECK_MEET(a, b, join) checkMeet(a, b, join, __FILE__, __LINE__)
+
+  void
+  checkLess(const Element& a, const Element& b, const char* file, int line) {
+    testing::ScopedTrace trace(file, line, "check less");
+    EXPECT_EQ(lattice.compare(a, b), analysis::LESS);
+    EXPECT_EQ(lattice.compare(b, a), analysis::GREATER);
+  }
+#define CHECK_LESS(a, b) checkLess(a, b, __FILE__, __LINE__)
+
+  void
+  checkGreater(const Element& a, const Element& b, const char* file, int line) {
+    testing::ScopedTrace trace(file, line, "check greater");
+    EXPECT_EQ(lattice.compare(a, b), analysis::GREATER);
+    EXPECT_EQ(lattice.compare(b, a), analysis::LESS);
+  }
+#define CHECK_GREATER(a, b) checkGreater(a, b, __FILE__, __LINE__)
+
+  void checkUnrelated(const Element& a,
+                      const Element& b,
+                      const char* file,
+                      int line) {
+    testing::ScopedTrace trace(file, line, "check unrelated");
+    EXPECT_EQ(lattice.compare(a, b), analysis::NO_RELATION);
+    EXPECT_EQ(lattice.compare(b, a), analysis::NO_RELATION);
+  }
+#define CHECK_UNRELATED(a, b) checkUnrelated(a, b, __FILE__, __LINE__)
+
+  void
+  checkEqual(const Element& a, const Element& b, const char* file, int line) {
+    testing::ScopedTrace trace(file, line, "check equal");
+    EXPECT_EQ(lattice.compare(a, b), analysis::EQUAL);
+    EXPECT_EQ(lattice.compare(b, a), analysis::EQUAL);
+  }
+#define CHECK_EQUAL(a, b) checkEqual(a, b, __FILE__, __LINE__)
+
+  void checkPair(const Element& a,
+                 const Element& b,
+                 const Element& join,
+                 const Element& meet,
+                 const char* file,
+                 int line) {
+    testing::ScopedTrace trace(file, line, "check pair");
+    CHECK_JOIN(a, b, join);
+    CHECK_MEET(a, b, meet);
+    switch (lattice.compare(a, b)) {
+      case analysis::NO_RELATION:
+        CHECK_UNRELATED(a, b);
```

Comment:
This is identical to what the switch is branching on, i.e., this adds nothing? It reads as
```js
if (x == 1) {
  assert(x == 1);
} else ..
```